### PR TITLE
Document underscore lifetime elision

### PR DESCRIPTION
**Pull request by matthewjasper**
cc #278 There don't seem to be any cases where `'_` could be used outside of this chapter.

`'_` can also be used for inferred lifetimes, but I don't think there is any documentation for them yet.

Also uses dyn trait in the elision chapter (cc #279)

### CODE REVIEW

File: src/lifetime-elision.md [hunk 1]
```diff
@@ -7,8 +7,10 @@ compiler can infer a sensible default choice.
 
 In order to make common patterns more ergonomic, Rust allows lifetimes to be
 *elided* in [function item], [function pointer] and [closure trait] signatures.
-The following rules are used to infer lifetime parameters for elided lifetimes.
-It is an error to elide lifetime parameters that cannot be inferred.
+An elided lifetime can be specified either by omitting it, or by using `'_`.
+For lifetimes in paths, using `'_` is preferred. The following rules are used
```
Comment (alercah) [inline src/lifetime-elision.md]: A few things here:

- What is a lifetime in a path? Do you mean as a parameter?
- Should probably mention that omitting a lifetime in a trait object type is equivalent to `'static`, not `'_`.

Comment (alercah) [reply src/lifetime-elision.md]: Also, is `'_` not preferred for references as well? I thought omitting lifetimes was going to be deprecated.

**Comment by matthewjasper** [comment]
Comments addressed.

> Also, is '_ not preferred for references as well? I thought omitting lifetimes was going to be deprecated.

The [RFC](http://rust-lang.github.io/rfcs/2115-argument-lifetimes.html#the-wildcard-lifetime) says

> When referring to a type (other than `&`/`&mut`) that requires lifetime arguments, it is deprecated to leave off those parameters.

**Comment by alercah** [comment]
Ah, my bad, excellent.

File: src/lifetime-elision.md [hunk 2]
```diff
@@ -41,22 +46,27 @@ fn get_mut<'a>(&'a mut self) -> &'a mut T;              // expanded
...
 
 ## Default trait object lifetimes
 
 The assumed lifetime of references held by a [trait object] is called its
 _default object lifetime bound_. These were defined in [RFC 599] and amended in
-[RFC 1156]. Default object lifetime bounds are used instead of the lifetime
-parameter elision rules defined above.
+[RFC 1156].
+
+> These default object lifetime bounds are used instead of the lifetime
```
Comment (Havvy) [inline src/lifetime-elision.md]: Is this supposed to be a Note?

Comment (matthewjasper) [reply src/lifetime-elision.md]: Well, it was intentional. I've changed it back to normal prose now.

File: src/lifetime-elision.md [hunk 3]
```diff
@@ -5,10 +5,14 @@ compiler can infer a sensible default choice.
 
 ## Lifetime elision in functions
 
-In order to make common patterns more ergonomic, Rust allows lifetimes to be
-*elided* in [function item], [function pointer] and [closure trait] signatures.
-The following rules are used to infer lifetime parameters for elided lifetimes.
-It is an error to elide lifetime parameters that cannot be inferred.
+In order to make common patterns more ergonomic, Rust allows lifetime argument
```
Comment (Havvy) [inline src/lifetime-elision.md]: Can you remove the "Rust allows" here?

**Pull request merged**